### ISSUE TITLE
Finish the multi-paragraph reading comprehension model

### ISSUE THREAD

**Issue by MaxMotovilov** [allenai/allennlp#1408]
Trying to determine the best match by running multiple paragraphs against the same query and getting inconclusive results. The output layer uses logit scores to find the matching span, but neither sum of the logits nor the product of the probabilities for span ends appear to correlate with the quality of the match when I vary the paragraph text slightly. Can I obtain a "normalized" (i.e. comparable across different paragraphs with constant query) measure of the prediction quality from the model output somehow?

The intended application is MC-aided search through large documents treated as a collection of sections, running the sections through BiDAF one at a time and comparing results.

Update - #1265 seems directly related. So this is more of a question of what's changed recently in that respect.

**Comment by joelgrus** [comment]
you may be interested in this model: https://arxiv.org/abs/1710.10723

I have a WIP PR #804 to get this into AllenNLP, hope to merge it soon

**Comment by MaxMotovilov** [comment]
Thanks! I might get to trying your code next week - will update on the outcome.

**Comment by MaxMotovilov** [comment]
@joelgrus - trying it, ran squarely into a problem with using multiple GPUs (can't finish training on a sensible subset in reasonable time on a single GPU). 

https://github.com/joelgrus/allennlp/blob/65bff0ac303cb601b6df1b4433de536ef8498494/allennlp/models/reading_comprehension/document_qa.py#L314

Here you seem to be assuming that the size of metadata is the same as the size of "batch" dimensions in the tensors. It isn't: `data_parallel()` distributes batches across GPUs by splitting the tensors so each call to `forward()` gets subsampled tensors (i.e. `batch_size` is something like 2 or 3 instead of 20 when running on 8 GPUs). For whatever reason, the metadata is not subsampled; I am a relative newbie in PyTorch and don't know the answer. What's obvious however is that correlating metadata dictionaries with tensor indices requires an extra map that reflects subsampling performed on tensors by PyTorch -- so that the code can do something like `metadata[ batch_subsample_map[i] ]` when iterating over tensor indices.

Any ideas how I can try and sort this out? I'll keep digging into PyTorch and AllenNLP code, but as I said, my experience with either is severely limited...

**Comment by murphp15** [comment]
@joelgrus @matt-gardner Are you guys at allennlp researching a solution to squad 2.0 dataset? 
This would be a really good solution to this confidence score issue.

**Comment by matt-gardner** [comment]
@murphp15, the initial baseline for SQuAD 2.0 was the model in #804. That model is definitely a good starting point for trying to solve the confidence score issue, and SQuAD 2.0. But no, none of us are actively working on this problem right now.

**Comment by matt-gardner** [comment]
We decided not to pursue this.  That model is pretty stale now, anyway.  If someone wants to contribute multi-paragraph reading comprehension work to AllenNLP, we'd be happy to have it, but this issue isn't doing any good by staying open.
